### PR TITLE
:sparkles: Upgrade to Kubernetes 1.28

### PR DESCRIPTION
**Pull request by guettli**
**What this PR does / why we need it**:

Upgrade to Kubernetes 1.28

Based on https://github.com/syself/cluster-api-provider-hetzner/pull/1015

### CODE REVIEW

File: pkg/services/baremetal/host/host.go [hunk 1]
```diff
@@ -1290,7 +1297,8 @@ func (s *Service) actionEnsureProvisioned() (ar actionResult) {
 		if err != nil {
 			if errors.Is(err, errUnexpectedHostName) {
 				// One possible reason: The machine gets used by a second wl-cluster
-				record.Warn(s.scope.HetznerBareMetalHost, "UnexpectedHostName", err.Error())
+				record.Warn(s.scope.HetznerBareMetalHost, "UnexpectedHostName",
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:1300]: warnf instead of warn with sprintf

File: pkg/services/baremetal/host/host.go [hunk 2]
```diff
@@ -1501,7 +1511,8 @@ func (s *Service) actionProvisioned() actionResult {
 			if err != nil {
 				if errors.Is(err, errUnexpectedHostName) {
 					// One possible reason: The machine gets used by a second wl-cluster
-					record.Warn(s.scope.HetznerBareMetalHost, "UnexpectedHostName", err.Error())
+					record.Warn(s.scope.HetznerBareMetalHost, "UnexpectedHostName",
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:1514]: warnf

File: Tiltfile [hunk 3]
```diff
@@ -33,7 +33,7 @@ settings = {
         "HCLOUD_REGION": "fsn1",
         "CONTROL_PLANE_MACHINE_COUNT": "3",
         "WORKER_MACHINE_COUNT": "3",
-        "KUBERNETES_VERSION": "v1.25.2",
+        "KUBERNETES_VERSION": "v1.27.3",
```
Comment (janiskemper) [inline Tiltfile:36]: is this correct?

File: controllers/hetznerbaremetalhost_controller_test.go [hunk 4]
```diff
@@ -243,7 +243,7 @@ var _ = Describe("HetznerBareMetalHostReconciler", func() {
 			By("making sure the it has been deleted")
 			Eventually(func() bool {
 				return apierrors.IsNotFound(testEnv.Get(ctx, key, host))
-			}, timeout, time.Second).Should(BeTrue())
+			}, 10*time.Second, time.Second).Should(BeTrue())
```
Comment (janiskemper) [inline controllers/hetznerbaremetalhost_controller_test.go:246]: We have talked about this already. We should find out the true reason why this is flaky. 10 seconds won't help

File: pkg/services/baremetal/host/host.go [hunk 5]
```diff
@@ -45,7 +45,7 @@ import (
 )
 
 const (
-	rebootWaitTime       time.Duration = 15 * time.Second
+	rebootWaitTime       time.Duration = 30 * time.Second
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:48]: are you sure that we want to increase this? This is about the time the server needs to start the reboot. AFAIK this should be done in a few seconds

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: Tiltfile (hunk 3 above)
Comment (kranurag7) [reply Tiltfile:36]: Yes, the stable version as per [kubernetes](https://www.downloadkubernetes.com/) for 1.27 is 1.27.8, but we don't have the kindest/node image with that tag. 
I used the latest tag that was present for 1.27. 
Ref: https://hub.docker.com/r/kindest/node/tags?page=1&name=v1.27

File: pkg/services/baremetal/host/host.go (hunk 1 above)
Comment (guettli) [reply pkg/services/baremetal/host/host.go:1300]: fixed on next push.

File: pkg/services/baremetal/host/host.go (hunk 2 above)
Comment (guettli) [reply pkg/services/baremetal/host/host.go:1514]: fixed on next push.

File: controllers/hetznerbaremetalhost_controller_test.go (hunk 4 above)
Comment (guettli) [reply controllers/hetznerbaremetalhost_controller_test.go:246]: @janiskemper I agree. I could not reproduce it. On my machine, the test was successful in 1s. I will use the old timeout again.

File: main.go [hunk 6]
```diff
@@ -98,16 +98,17 @@ func main() {
...
+		HealthProbeBindAddress: probeAddr,
+		WebhookServer: webhook.NewServer(webhook.Options{
+			Port: 9443,
+		}),
 		LeaderElection:                enableLeaderElection,
 		LeaderElectionID:              "hetzner.cluster.x-k8s.io",
 		LeaderElectionNamespace:       leaderElectionNamespace,
 		LeaderElectionResourceLock:    "leases",
 		LeaderElectionReleaseOnCancel: true,
-		Namespace:                     watchNamespace,
```
Comment (janiskemper) [inline main.go:110]: why do you remove this? CAPI still uses it https://github.com/kubernetes-sigs/cluster-api/blob/main/main.go#L301

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: pkg/services/baremetal/host/host.go (hunk 5 above)
Comment (guettli) [reply pkg/services/baremetal/host/host.go]: ok, I will revert my change.

File: controllers/hetznerbaremetalhost_controller_test.go [hunk 7]
```diff
@@ -336,13 +336,16 @@ var _ = Describe("HetznerBareMetalHostReconciler", func() {
 				}, timeout, time.Second).Should(BeNil())
 
 				Eventually(func() bool {
-					// TODO: Add logging to trace flaky unit-test.
 					if err := testEnv.Get(ctx, key, host); err != nil {
+						testEnv.GetLogger().Info("......... reaches the state image installing. Get failed", "err", err)
```
[the remark this answers is not in the document]
Comment (guettli) [reply controllers/hetznerbaremetalhost_controller_test.go]: done.

File: main.go (hunk 6 above)
Comment (guettli) [reply main.go:110]: I get:

> unknown field Namespace in struct literal of type manager.Options

From controller-runtime release page:

> cache.Options.Namespace has been removed in favor of cache.Options.Namespaces

Comment (janiskemper) [reply main.go:110]: yes just check out the CAPI code!

Comment (guettli) [reply main.go:110]: According to the [contract](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/support-multiple-instances#contract), it is required. I added it again.

**Review by janiskemper — APPROVED** [review]

**Pull request merged**